thrust/system/cuda/detail/core/util.h

```diff
@@ -615,11 +616,55 @@ namespace core {
   };
 
   THRUST_RUNTIME_FUNCTION
-  inline cuda_optional<int> get_ptx_version()
+  inline int get_ptx_version()
   {
+    const int current_device = cub::CurrentDevice();
+    if (current_device < 0)
+    {
+      cuda_cub::throw_on_error(cudaErrorNoDevice, "No GPU is available\n");
+    }
```

Comment:
I think this call to `cub::CurrentDevice()` and throwing the exception if it fails should be moved just inside the `if (cub::PtxVersion(ptx_version) != cudaSuccess)` block.  That results in one less CUDA runtime call when we are on the happy path of having the right GPU.  This suggestion assumes that `cub::PtxVersion` will return something other than success when there is no GPU.

Reply:
I've missed that optimization, thanks!